src/components/form/form.jsx

```diff
@@ -15,27 +17,35 @@ const Form = React.forwardRef((props, forwardedRef) => {
   const [dirty, setDirty] = useState(false);
   const [valid, setValid] = useState(true);
 
+  const onDebouncedSubmit = useCallback(debounce(() => {
+    if (ref.current.checkValidity() && ref.current.dirty) {
+      props.onSubmit({
+        target: ref.current,
+        data: props.refs.reduce((data, controlRef) => {
+          if (controlRef.current) {
+            return {
+              ...data,
+              [controlRef.current.name]: controlRef.current,
+            };
+          }
+
+          return data;
+        }, {}),
+      });
+    }
+  }, 300, { leading: !props.autoSave, trailing: props.autoSave }), [props.onSubmit]);
```

Comment:
Whoever uses this will have to be careful. If their `onSubmit` prop is generated every time their component re-renders then we will be creating new debounce functions here and will lose the debounciness of it all. One possible solution would be to do something like `const memoizedOnSubmit = useCallback(() => props.onSubmit(), []). I guess at that point just removing `props.onSubmit` from the `onDebouncedSubmit` dependency array would get you the same thing. There would be subtle bugs if the `props.onSubmit` function relied on closures.

Reply:
I had that originally but found it was buggy since the `onSubmit` callback _can_ rely on closures. 

Instead, calculating a new debounce when the closure changes should be fine. 

1. Create a reference to `onSubmit` (with a dependency on the component's state)
1. Parent passes a reference to `Form`, `Form` caches the debounced function
1. Form invokes the debounce function because the user is typing
1. The reference to `onSubmit` is not yet called, the parent component has no reason to believe it needs to change _anything_
1. Enough time passes by, Form invokes the reference to `onSubmit`. 
1. The parent updates _something_. It most likely created a new closure for the `onSubmit`
1. Parent passes a reference to the new `onSubmit`. `Form` caches the debounced function.
1. ...

There is a low likelihood the parent will re-render its children while the debounce function is being debounced. The user is in the middle of typing stuff so it doesn't make sense that the parent component will re-render.

However, if for some reason the parent does re-render then it's up to the implementation to memoize their own `onSubmit` that will not cause new debounced function. 

Does that make sense? 

Maybe a good way to think about it is that it's buggy either way? At least the onus is on the parent component when we debounce new `onSubmit`s

Reply:
TODO: Add documentation about the gotchas.

Reply:
Keeping the responsibility on the parent component does seem like the lesser of the two evils. Keeps this component as composable(and its behavior reliable) as possible.